test/scripts/parallel-local/main.js

```diff
@@ -0,0 +1,234 @@
+ */
+const initializeWorkers = () => {
+    const workers = [];
+
+    for (let i = 0; i < numWorkers; i++) {
+        const port = basePort + i;
+        const projectName = `worker-${i}`;
+        const worker = fork(path.resolve(__dirname, 'test-runner.js'), [port.toString(), projectName], {
+            stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
+        });
+
+        setupWorkerListeners(worker, port, projectName);
+        assignTestToWorker(worker, projectName);
+        workers.push(worker);
+    }
+};
+
+/**
+ * Sets up event listeners for a worker process.
+ * @param {ChildProcess} worker - The worker process.
+ * @param {number} port - The port number assigned to the worker.
+ * @param {string} projectName - The project name assigned to the worker.
+ * @returns {void} Does not return a value.
+ */
+const setupWorkerListeners = (worker, port, projectName) => {
+    worker.stdout.on('data', (data) => console.log(`${projectName}: ${data.toString()}`));
+    worker.stderr.on('data', (data) => console.error(`${projectName} Error: ${data.toString()}`));
+
+    worker.on('message', (msg) => handleWorkerMessage(msg, worker, projectName));
+    worker.on('exit', (code) => handleWorkerExit(code, projectName));
+};
+
+/**
+ * Handles messages from worker processes, potentially assigning new tests.
+ * @param {string} msg - The message from the worker.
+ * @param {ChildProcess} worker - The worker process.
+ * @param {string} projectName - The project name.
+ * @returns {void} Does not return a value.
+ */
+const handleWorkerMessage = (msg, worker, projectName) => {
+    if (msg === 'request_next_test') {
+        assignTestToWorker(worker, projectName);
+    }
+};
+
+/**
+ * Assigns a test to a worker and logs the activity.
+ * @param {ChildProcess} worker - The worker process.
+ * @param {string} projectName - The project name.
+ * @returns {void} Does not return a value.
+ */
+const assignTestToWorker = (worker, projectName) => {
+    if (remainingTests.length > 0) {
+        const test = remainingTests.pop();
+        console.log(`${projectName} starting new suite: ${test} (${testSuites.length - remainingTests.length}/${testSuites.length})`);
+        worker.send({ test, projectName });
+    } else {
+        console.log(`${projectName} found no more tests...`);
+        worker.send('no_more_tests');
+        workersCompleted.add(projectName);
+    }
+};
+
+/**
+ * Logs the exit of worker processes and checks for completion.
+ * @param {number} code - The exit code of the worker process.
+ * @param {string} projectName - The project name.
+ * @returns {void} Does not return a value.
+ */
+const handleWorkerExit = (code, projectName) => {
+    if (code !== 0) {
+        console.error(`Worker ${projectName} exited with code ${code}`);
+    } else {
+        console.log(`Worker ${projectName} completed successfully`);
+    }
+    workersExited.add(projectName);
+    if (workersCompleted.size === numWorkers && workersExited.size === numWorkers) {
```

Comment:
What if a worker crashed? I suppose that workerExited condition will be fine, but not the workerCompleted ?

Reply:
Done, also extracted to a function in case we want to do something in case of a crash in the future